### PR TITLE
Add Spring Boot Starter for Easy Rules

### PR DESCRIPTION
**Pull request by andersonkyle**
A Spring Boot Starter for Easy Rules to make it even _easier_ to get up and running for Spring Boot projects.

### CODE REVIEW

**Comment by benas** [comment]
Hi Kyle,

Fantastic! This is awesome 👍 thank you very much.
Indeed, it's easier to get started with this starter.

Please correct me if I'm wrong, I'm not spring boot expert.
If I understand well, I can start like this:

* create a blank maven project
* add spring boot easy rules starter in the pom
* Create my rules (they will be scanned by the EasyRulesAutoConfiguration class)
* Write my easy rules app, For example:

```java
@SpringBootApplication
@EnableEasyRules // I know, this annotation does not exist yet. What it the equivalent in your PR?
public class EasyRulesApplication {

	@Autowired
	private RulesEngine rulesEngine;
	
	public static void main(String[] args) throws Exception {
		rulesEngine.fireRules();
	}

}
```

Is this correct? I quickly saw your changes but not (yet) in details.

Many thanks upfront
Best regards
Mahmoud

**Comment by andersonkyle** [comment]
Hi Mahmoud,

Your understanding is correct.  `EasyRulesAutoConfiguration` provides the `RulesEngineFactoryBean` which in turn generates a `RulesEngine `bean already populated with all of your Rules.

We don't need an `@EnableEasyRules` annotation because we're using conditions on the autoconfiguration to determine whether it is enabled or not.
```java
@ConditionalOnClass(RulesEngineFactoryBean.class)
@ConditionalOnMissingBean(RulesEngineFactoryBean.class)
```
EasyRules will only be autoconfigured if the `RulesEngineFactoryBean `class is present on the classpath, and if there isn't a `RulesEngineFactoryBean `already defined.

File: easyrules-spring-boot-starter/src/test/java/org/easyrules/spring/boot/autoconfigure/EasyRulesAutoConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        assertEquals(1, rulesEngine.getRules().size());
+    }
+
+    @Test
+    public void springRuleAnnotationTest() {
+        load(SpringRuleAnnotationConfiguration.class);
+        RulesEngine rulesEngine = this.context.getBean(RulesEngine.class);
+        assertEquals(1, rulesEngine.getRules().size());
+    }
+
```
Comment (benas) [inline easyrules-spring-boot-starter/src/test/java/org/easyrules/spring/boot/autoconfigure/EasyRulesAutoConfigurationTest.java]: I think we can add a test case when we have multiple configuration classes:

```java
@Test
public void multipleRulesTest() {
    load(SpringRuleAnnotationConfiguration.class);
    load(RuleComponentAnnotationConfiguration.class);
    load(RuleComponentAnnotationConfiguration.class);
    // TODO adapt the load method to take multiple configuration classes
    RulesEngine rulesEngine = this.context.getBean(RulesEngine.class);
    assertEquals(3, rulesEngine.getRules().size());
}
```

What do you think?

**Comment by benas** [comment]
Ok, great. It's even easier than I thought 😄 

There still a point I don't understand enough. It's about rules registration.  Let's say this is my spring boot app (without EnableEasyRules annotation):

```java
@SpringBootApplication
public class EasyRulesApplication {

	@Autowired
	private RulesEngine rulesEngine;
	
	public static void main(String[] args) throws Exception {
		rulesEngine.fireRules();
	}

}
```

From the `EasyRulesAutoConfiguration#rulesEngineFactoryBean` I see that rules should be managed by Spring (which I understand) and they are taken from a spring application context. My question is: where do I tell spring boot which application context to load rules from (be it xml or annotation app context) ?

**Comment by andersonkyle** [comment]
Most Spring Boot apps call [this](https://github.com/spring-projects/spring-boot/blob/master/spring-boot/src/main/java/org/springframework/boot/SpringApplication.java#L1244-L1253) static helper method to create and run a new `ApplicationContext`

```java
@SpringBootApplication
public class EasyRulesApplication {

	@Autowired
	private static RulesEngine rulesEngine;
	
	public static void main(String[] args) throws Exception {
                SpringApplication.run(EasyRulesApplication.class, args);
		rulesEngine.fireRules();
	}

}
```
The context will automatically include all your Beans, Components, Services, etc. via `@Configuration` and `@ComponentScan`.  You can certainly override this behavior and do it yourself but I'd say the majority of Spring Boot apps leverage this `autoconfiguration` that takes care of a lot of the Spring Framework plumbing.

File: easyrules-spring-boot-starter/src/test/java/org/easyrules/spring/boot/autoconfigure/EasyRulesAutoConfigurationTest.java (hunk 1 above)
Comment (andersonkyle) [inline easyrules-spring-boot-starter/src/test/java/org/easyrules/spring/boot/autoconfigure/EasyRulesAutoConfigurationTest.java]: Absolutely.  PR has been updated.

**Comment by benas** [comment]
Thank you for updating the PR. This starter is great! making something easy easier 😉 

I'm merging your changes.

Thank you!

**Comment by cgonul** [comment]
Hi there;

I can not get the spring annoteted rules when I use the spring boot starter.
```

	@Autowired
	private ApplicationContext applicationContext;

	public void coin() {
		RulesEngine rulesEngine = RulesEngineBuilder.aNewRulesEngine().build();
		rulesEngine.getParameters().setSkipOnFirstFailedRule(true);
		rulesEngine.registerRule(new DummyRule());
		rulesEngine.registerRule(applicationContext.getBean(SpringManagedRule.class));
		rulesEngine.fireRules();
	}

The dummy rule is successfully registering however SpringManagedRule is not registering.
```

import java.util.List;

import org.apache.commons.collections.CollectionUtils;
import org.easyrules.annotation.Action;
import org.easyrules.annotation.Condition;
import org.easyrules.spring.SpringRule;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;

import com.netas.aydes.geologicalsurvey.dao.dto.impl.GsBuildingFormDTO;
import com.netas.aydes.geologicalsurvey.service.GsBuildingFormService;

@SpringRule
public class SpringManagedRule {

	private static final Logger logger = LoggerFactory.getLogger(DummyRule.class);

	@Autowired
	private Service service;

	@Condition
	public boolean when() {
		return !service.isNotEmpty(0);
	}

	@Action
	public void then() {
		logger.info("Hey, I'm managed by Spring");
	}
}
```
```

Do you have any idea what I am doing wrong. I am using 2.4.1.SNAPSHOT. Your tutorials works perfectly. But I can not find a way of binding spring rule when I use spring-boot.

Thanks.

**Comment by benas** [comment]
Hi Kyle,

I am testing this new starter to prepare release v2.5 and getting some errors.
Here is the project I'm using for test: https://gist.github.com/benas/2d42444a627c3eca1e3b5eff0897d856

When I run the app, I get the error in `error1.log`.
Looks like spring boot starter web is missing. I tried to add in the starter, rebuild the starter and rerun the app. And now I get the error in `error2.log`.

Do you have an idea on what's happening? Or I forgot to add something?
I'm expecting to only add the starter dependency in my project and it should.

Your help is much appreciated!
Kind regards
Mahmoud

**Comment by benas** [comment]
I found the problem. Looks like putting classes in the default package is the cause (see issue 986 in spring boot project). So I've put classes in a `org.test` package and now there is no more dependencies problem. But I get a NullPointerException on the rules engine:

```
Caused by: java.lang.NullPointerException
	at org.test.EasyRulesApp.main(EasyRulesApp.java:16)
	... 6 more
```

**Comment by andersonkyle** [comment]
Hi Mahmoud,

A quick suggestion is to put the spring-boot-starter-parent POM as the Parent of your POM.

Also try using the spring-boot-maven-plugin instead of the exec plugin. 

Run your app with mvn spring-boot:run

I'm traveling these next few days and am away from my laptop but as soon as I get back I can give you a more thoughtful answer. 

Regards, 

Kyle

**Comment by andersonkyle** [comment]
One more suggestion, if the above doesn't work, add spring-boot-starter as a dependency to your app and then run it again.

**Comment by benas** [comment]
Ok great, thank you for these suggestions!

> I'm traveling these next few days and am away from my laptop but as soon as I get back I can give you a more thoughtful answer.

No problem, have a safe travel!

**Comment by benas** [comment]
I found the problem. By adding the spring-boot-starter as a dependency, more logs appear now and I see the following : 

```
2017-05-14 23:19:15.600  WARN 14171 --- [           main] f.a.AutowiredAnnotationBeanPostProcessor :
Autowired annotation is not supported on static fields: 
private static org.easyrules.api.RulesEngine org.test.EasyRulesApp.rulesEngine
```

So I updated the app to this:

```java
@SpringBootApplication
public class EasyRulesApp {

    public static void main(String[] args) throws Exception {
        ConfigurableApplicationContext applicationContext = SpringApplication.run(EasyRulesApp.class, args);
        RulesEngine rulesEngine = applicationContext.getBean(RulesEngine.class);
        rulesEngine.fireRules();
    }
}
```

and now it is working fine.

Thank you for your helpful suggestions!

**Pull request merged**